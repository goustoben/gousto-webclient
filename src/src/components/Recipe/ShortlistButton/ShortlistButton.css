@value Bluecheese, BluecheeseLighter, White, Pomegranate, Mackerel, Cauliflower, DarkCauliflower from 'styles/colors.css';
@value headingFontWeightBold from 'styles/typography.css';

.shortlistButton {
  composes: marginSMTop from 'sass-loader!styles/spacing.scss';

  flex: 1;
  
  margin-right: 0.5rem;
  padding: 0;
  border: 1px solid Bluecheese;
  border-radius: 3px;
  text-align: center;
  position: relative;
  background: White;
}

.shortlistButton:disabled {
  background: Mackerel;
  border-color: Bluecheese;
}

.shortlistButton:disabled .shortlistText {
  color: Bluecheese;
}

.shortlistButton.shortlisted {
  background: BluecheeseLighter;
}

.heartIcon {
  height: 1rem;
  width: 1.125rem;
  display: inline-block;
  vertical-align: middle;
  margin-bottom: 0.125rem;
  margin-right: 0.5rem;
}

.shortlistText {
  text-transform: uppercase;
  color: Bluecheese;
  font-weight: headingFontWeightBold;
  vertical-align: middle;
}
