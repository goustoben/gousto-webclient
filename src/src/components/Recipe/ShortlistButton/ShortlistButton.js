import React from 'react'
import PropTypes from 'prop-types'
import Svg from 'Svg'
import classnames from 'classnames'
import { ShortlistTutorial } from 'routes/Menu/ShortlistTutorial'
import css from './ShortlistButton.css'

class ShortlistButton extends React.PureComponent {

  static propTypes = {
    shortlistLimitReached: PropTypes.bool,
    recipeInShortlist: PropTypes.bool,
    addToShortlist: PropTypes.func,
    removeFromShortlist: PropTypes.func,
    menuBrowseCTAVisibilityChange: PropTypes.func,
    stock: PropTypes.number,
    id: PropTypes.string,
    position: PropTypes.number,
    showShortListTutorial: PropTypes.bool,
    shortlistTutorialStep1Viewed: PropTypes.bool,
    incrementTutorialViewed: PropTypes.func,
    tutorialTracking: PropTypes.func,
    view: PropTypes.string,
    notInStock: PropTypes.bool,
  }

  static defaultProps = {
    shortlistLimitReached: false,
    recipeInShortlist: false,
    addToShortlist: () => { },
    removeFromShortlist: () => { },
    menuBrowseCTAVisibilityChange: () => { },
    stock: 0,
    id: '',
    position: 0,
    showShortListTutorial: false,
    shortlistTutorialStep1Viewed: false,
    view: 'grid',
    notInStock: false
  }

  closeTutorialStep1 = () => {
    const { incrementTutorialViewed, tutorialTracking } = this.props
    const tutorialName = 'shortlistStep1'
    incrementTutorialViewed(tutorialName)
    tutorialTracking(tutorialName, 0, true)
  }

  onShortlistClick = () => {
    const { removeFromShortlist, recipeInShortlist, id, shortlistLimitReached, view } = this.props

    if (recipeInShortlist) {
      removeFromShortlist(id, { view })
    } else if (!shortlistLimitReached) {
      this.handleAdd()
    }
  }

  handleAdd = () => {
    const { addToShortlist, menuBrowseCTAVisibilityChange, id, position, stock, shortlistTutorialStep1Viewed, view } = this.props
    const recipeInfo = { position, view }

    if (stock !== null) {
      if (!shortlistTutorialStep1Viewed) {
        this.closeTutorialStep1()
      }
      addToShortlist(id, false, recipeInfo)
    } else {
      menuBrowseCTAVisibilityChange(true)
    }
  }

  render() {
    const { shortlistLimitReached, recipeInShortlist, showShortListTutorial, notInStock } = this.props
    const shortlistButtonText = (recipeInShortlist ? 'Shortlisted' : 'Add to shortlist')
    const classes = classnames(
      css.shortlistButton,
      (recipeInShortlist ? css.shortlisted : null)
    )

    if (notInStock) {
      return null
    }

    return (
      <button type='button' disabled={shortlistLimitReached && !recipeInShortlist} data-slug='checkmark' onClick={this.onShortlistClick} className={classes}>
        {shortlistButtonText}
        {recipeInShortlist && <Svg fileName={'icon_tick_checkmark'} className={css.checkmark} /> }
        {showShortListTutorial && <ShortlistTutorial />}
      </button>
    )
  }
}

export { ShortlistButton }
