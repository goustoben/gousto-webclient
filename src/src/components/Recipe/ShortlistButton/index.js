export { ShortlistButtonContainer as ShortlistButton } from './ShortlistButtonContainer'
