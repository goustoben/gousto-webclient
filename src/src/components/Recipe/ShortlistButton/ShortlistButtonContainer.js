import { connect } from 'react-redux'
import { getShortlist } from 'selectors/features'
import { getShortlistLimitReached } from 'selectors/basket'
import { shortlistRecipeAdd, shortlistRecipeRemove } from 'actions/shortlist'
import { menuBrowseCTAVisibilityChange } from 'actions/menu'
import { ShortlistButton } from './ShortlistButton'

function mapStateToProps(state) {

  return {
    showShortlistButton: getShortlist(state),
    shortlistLimitReached: getShortlistLimitReached(state)
  }
}

const ShortlistButtonContainer = connect(mapStateToProps, {
  addToShortlist: shortlistRecipeAdd,
  removeFromShortlist: shortlistRecipeRemove,
  menuBrowseCTAVisibilityChange,
})(ShortlistButton)

export { ShortlistButtonContainer }
