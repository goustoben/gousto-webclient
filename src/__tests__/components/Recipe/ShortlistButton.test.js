import React from 'react'
import { shallow, mount } from 'enzyme'
import { ShortlistButton } from 'Recipe/ShortlistButton/ShortlistButton'

describe('<ShortlistButton />', () => {
  let wrapper
  const addToShortlistSpy = jest.fn()
  const removeFromShortlistSpy = jest.fn()
  const menuBrowseCTAVisibilityChangeSpy = jest.fn()
  const shortlistButtonProps = {
    shortlistLimitReached: false,
    addToShortlist: addToShortlistSpy,
    removeFromShortlist: removeFromShortlistSpy,
    menuBrowseCTAVisibilityChange: menuBrowseCTAVisibilityChangeSpy,
    stock: 1000,
    id: '1234',
    position: 1
  }

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('Render', () => {
    test('should show SVG when renders', () => {
      wrapper = shallow(<ShortlistButton {...shortlistButtonProps} />)

      expect(wrapper.find('Svg').length).toBe(1)
    })

    test('should show blue heart (deseleted) if recipeInShorlist is false', () => {
      wrapper = shallow(<ShortlistButton {...shortlistButtonProps} />)
      wrapper.setState({recipeInShortlist: false})

      expect(wrapper.find("[fileName='icon_shortlist_heart_deselected']").length).toBe(1)
      expect(wrapper.find('.blueHeartButton').length).toBe(1)
    })

    test('should show red heart (seleted) if recipeInShorlist is true', () => {
      wrapper = shallow(<ShortlistButton {...shortlistButtonProps} />)
      wrapper.setProps({recipeInShortlist: true})

      expect(wrapper.find("[fileName='icon_shortlist_heart_selected']").length).toBe(1)
      expect(wrapper.find('.redHeartButton').length).toBe(1)
    })

    test('should include defaultDetailView css if display is detailOverview', () => {
      wrapper = shallow(<ShortlistButton {...shortlistButtonProps} />)
      wrapper.setProps({display: 'detailOverlay'})

      expect(wrapper.find('.defaultDetailView').length).toBe(1)
    })
  })

  describe('onShortlistClick', () => {
    wrapper = mount(<ShortlistButton {...shortlistButtonProps }/>)

    test('should add recipe to shortlist if recipeInShortlist is false, shortlistLimitReached isnt reached and recipe is in stock - and change recipeInShortlist state', () => {
      wrapper.setProps({recipeInShortlist: false})
      wrapper.find('button').simulate('click')

      expect(addToShortlistSpy).toHaveBeenCalledTimes(1)
    })

    test('should call menuBrowseCTAVisibilityChange if stock is null', () => {
      wrapper.setProps({stock: null, recipeInShortlist: false})
      wrapper.find('button').simulate('click')

      expect(menuBrowseCTAVisibilityChangeSpy).toHaveBeenCalledTimes(1)
      expect(addToShortlistSpy).not.toHaveBeenCalled()
      expect(removeFromShortlistSpy).not.toHaveBeenCalled()
    })

    test('should NOT add recipe to shortlist if recipeInShortlist is false and shortlistLimitReached is reached', () => {
      wrapper.setProps({stock: 1000, shortlistLimitReached: true, recipeInShortlist: false})

      wrapper.find('button').simulate('click')

      expect(addToShortlistSpy).not.toHaveBeenCalled()
      expect(removeFromShortlistSpy).not.toHaveBeenCalled()
    })

    test('should remove recipe from shortlist if recipeInShortlist is true - and change recipeInShortlist state', () => {
      wrapper.setProps({recipeInShortlist: true})

      wrapper.find('button').simulate('click')

      expect(removeFromShortlistSpy).toHaveBeenCalledTimes(1)
      expect(addToShortlistSpy).not.toHaveBeenCalled()
    })
  })
})
