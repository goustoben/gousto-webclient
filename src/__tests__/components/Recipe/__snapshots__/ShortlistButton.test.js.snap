// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShortlistButton /> Render should render correctly if recipeInShortlist is false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShortlistButton
    addToShortlist={[MockFunction]}
    id="1234"
    incrementTutorialViewed={[MockFunction]}
    menuBrowseCTAVisibilityChange={[MockFunction]}
    notInStock={false}
    position={1}
    recipeInShortlist={false}
    removeFromShortlist={[MockFunction]}
    shortlistLimitReached={false}
    shortlistTutorialStep1Viewed={false}
    showShortListTutorial={false}
    stock={1000}
    tutorialTracking={[MockFunction]}
    view="grid"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Svg
          className="heartIcon"
          fileName="icon_shortlist_heart_deselected"
        />,
        false,
      ],
      "className": "shortlistButton",
      "data-slug": "heart",
      "disabled": false,
      "onClick": [Function],
      "onKeyPress": [Function],
      "tabIndex": 0,
      "type": "button",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "heartIcon",
          "fileName": "icon_shortlist_heart_deselected",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Svg
            className="heartIcon"
            fileName="icon_shortlist_heart_deselected"
          />,
          false,
        ],
        "className": "shortlistButton",
        "data-slug": "heart",
        "disabled": false,
        "onClick": [Function],
        "onKeyPress": [Function],
        "tabIndex": 0,
        "type": "button",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "heartIcon",
            "fileName": "icon_shortlist_heart_deselected",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<ShortlistButton /> Render should show red heart (seleted) if recipeInShorlist is true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShortlistButton
    addToShortlist={[MockFunction]}
    id="1234"
    incrementTutorialViewed={[MockFunction]}
    menuBrowseCTAVisibilityChange={[MockFunction]}
    notInStock={false}
    position={1}
    recipeInShortlist={true}
    removeFromShortlist={[MockFunction]}
    shortlistLimitReached={false}
    shortlistTutorialStep1Viewed={false}
    showShortListTutorial={false}
    stock={1000}
    tutorialTracking={[MockFunction]}
    view="grid"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Svg
          className="heartIcon"
          fileName="icon_shortlist_heart_selected"
        />,
        false,
      ],
      "className": "shortlistButton shortlisted",
      "data-slug": "heart",
      "disabled": false,
      "onClick": [Function],
      "onKeyPress": [Function],
      "tabIndex": 0,
      "type": "button",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "heartIcon",
          "fileName": "icon_shortlist_heart_selected",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Svg
            className="heartIcon"
            fileName="icon_shortlist_heart_selected"
          />,
          false,
        ],
        "className": "shortlistButton shortlisted",
        "data-slug": "heart",
        "disabled": false,
        "onClick": [Function],
        "onKeyPress": [Function],
        "tabIndex": 0,
        "type": "button",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "heartIcon",
            "fileName": "icon_shortlist_heart_selected",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
